include/trieste/source.h

```diff
@@ -71,46 +71,100 @@ namespace trieste
+        pos,
+        [](std::pair<size_t, size_t> elem, size_t pos_) {
+          return elem.first <= pos_;
+        });
+      // If we're at the beginning already, or we couldn't find a line after
+      // pos, then our current line should be fine. If lines is constructed
+      // correctly, it is on either the only line or the last line (and pos is
+      // on that line or beyond all lines).
+      if (it == lines.end() || (it != lines.begin() && it->first > pos))
+      {
+        --it;
+      }
 
-      if (it != lines.begin())
-        col -= *(it - 1) + 1;
+      size_t line = std::distance(lines.begin(), it);
+      size_t col = pos - it->first;
 
       return {line, col};
     }
 
     std::pair<size_t, size_t> linepos(size_t line) const
     {
-      // Lines are 0-indexed.
-      if (line > lines.size())
-        return {std::string::npos, 0};
-
-      size_t start = 0;
-      auto end = contents.size();
-
-      if (line > 0)
-        start = lines[line - 1] + 1;
-
-      if (line < lines.size())
-        end = lines[line];
+      // Special case: for out of range lines, index them at the end of our
+      // string and give them length 0. This will cause minimally-ugly
+      // misbehavior if there's an out of range error. Also, this gracefully
+      // handles the technically out of range situation where we ask for line 0
+      // of an empty string, which is what linecol(0) will give its caller in
+      // that case.
+
+      // Change note: this used to return std::string::npos when line was out of
+      // range by more than 1, but callers weren't checking for it so that case
+      // just caused things like .subtr(max_long) in practice. Best return an
+      // empty line with max idx, which will cause blank outputs. Otherwise,
+      // actually assert(line <= lines.size()) to crash here and not half way
+      // down the calling function.
+      if (line >= lines.size())
+      {
+        return {contents.size(), 0};
+      }
 
-      return {start, end - start};
+      return lines[line]; // already in {start, size} format
     }
 
   private:
     void find_lines()
     {
-      // Find the lines.
-      auto pos = contents.find('\n');
+      using namespace std::string_view_literals;
+      // Find the lines (accounting for cross-platform line ending issues).
+      // We store the size of the line to support linepos(line), so people can
+      // print exactly the line and not fragments of \r\n.
+      size_t cursor = 0;
+      auto try_match = [&](std::string_view part) -> bool {
+        if (std::string_view(contents).substr(cursor, part.size()) == part)
+        {
+          cursor += part.size();
+          return true;
+        }
+        else
+        {
+          return false;
+        }
+      };
+
+      size_t line_start = 0;
+      while (cursor < contents.size())
+      {
+        size_t last_pos = cursor;
+        if (try_match("\r\n"sv) || try_match("\n"sv) || try_match("\r"sv))
```

Comment:
Just a quibble but a line cannot end with carriage return. The only valid line endings are `\r\n` or `\n`. As unlikely as it may be, we do not want to be line breaking on `\r`.

Reply:
I saw someone on StackOverflow talking about an old Mac system that had `/r` as a line break... which upon further inspection was OS 9 and has been obsolete since 2001.

Fair enough, my bad for not checking that more thoroughly. Easy enough to remove.

Reply:
Removed.